web/api/query.go

```diff
@@ -40,22 +40,44 @@ func setAccessControlHeaders(w http.ResponseWriter) {
 	w.Header().Set("Access-Control-Expose-Headers", "Date")
 }
 
+func httpJSONError(w http.ResponseWriter, err error, code int) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(code)
+	fmt.Fprintln(w, ast.ErrorToJSON(err))
+}
+
+func parseTimestampOrNow(t string) (clientmodel.Timestamp, error) {
+	if t == "" {
+		return clientmodel.Now(), nil
+	}
+
+	tFloat, err := strconv.ParseFloat(t, 64)
+	if err != nil {
+		return 0, err
+	}
+	return clientmodel.TimestampFromUnixNano(int64(tFloat * float64(time.Second/time.Nanosecond))), nil
+}
+
+func parseDuration(d string) (time.Duration, error) {
+	dFloat, err := strconv.ParseFloat(d, 64)
+	if err != nil {
+		return 0, err
+	}
+	return time.Duration(dFloat * float64(time.Second/time.Nanosecond)), nil
+}
+
 // Query handles the /api/query endpoint.
 func (serv MetricsService) Query(w http.ResponseWriter, r *http.Request) {
 	setAccessControlHeaders(w)
+	w.Header().Set("Content-Type", "application/json")
 
 	params := httputils.GetQueryParams(r)
 	expr := params.Get("expr")
-	asText := params.Get("asText")
-
-	var format ast.OutputFormat
-	// BUG(julius): Use Content-Type negotiation.
-	if asText == "" {
-		format = ast.JSON
-		w.Header().Set("Content-Type", "application/json")
-	} else {
-		format = ast.Text
-		w.Header().Set("Content-Type", "text/plain")
+
+	timestamp, err := parseTimestampOrNow(params.Get("timestamp"))
+	if err != nil {
+		httpJSONError(w, fmt.Errorf("invalid query timestamp %s", err), http.StatusBadRequest)
```

Comment:
Yeah, `Set()` just overwrites whatever is there, while `Add()` appends. I decided to do it like this to make sure that the right header is always set, even though it's not technically needed right now, as we already set the header at the beginning of each API method.